Support for N-depth in module structure

splashx opened the pull request:
Users can define a subcommand in the format `<command>-<subcommand>`.
It expects the subcommand to be located in a folder under `commands/<command>/sub-command.py`

splashx committed:
add support hiphenized sub-commands

splashx committed:
add support for subcomand in format <command>-<subcommand>

splashx committed:
Add sub-command explanation in README

splashx committed:
add multi-level support using dash as field separator

jesselang commented on README.md:
```diff
@@ -27,16 +27,52 @@ directory layout:
+           |-- <level1>
+                |
+                |-- __init__.py
+                |-- <level2>.py
```
Can we have both a `<level2>.py` file and `<level2>` directory, and show how commands could be loaded from each?

splashx replied:
Done in 6d7f343da2aa4b870b5736733fc37e0098ca6adc - in fact `<level2>` was done but wrongly added `.py`. Added `<command>` in `level1` to show that's also possible.

jesselang commented on bin/cog-command:
```diff
@@ -5,8 +5,15 @@ import os
+# pycog3 <= 0.1.27 throws exception
+path = command_name.replace("-",".")
+full_path = "%s.commands.%s" % (bundle_name, path)
+class_name = command_name.split("-")[-1].capitalize()
```
This is a nice direction. I wonder if we could put some testing around this to help ensure that we're getting the desired results with all variations of project organizations. I'd feel good about merging this if we can prove that it won't break backward compatibility.

jesselang commented on setup.py:
```diff
@@ -3,13 +3,13 @@
-    version = "0.1.27",
-    scripts = ["bin/cog-command"],
+    version = "0.1.28",
+    scripts = ["bin/cog-command",],
```
Unnecessary comma?

splashx replied:
Cosmetic - this is also an acceptable python syntax.  Removed.

splashx committed:
fixes level2 example and add command at level1

splashx committed:
remove comma - cosmetic change

splashx committed:
Adds support for multi-level command using hiphen as command separator (#1)

* add testbundle, add tests/, add .travis-ci.yml

* rename .travis-ci.yml to .travis.yml

* remove unused requirements, fix tests

* add test test_command

* consolidate function names in test_command

* remove unused dependencies

* Change license

* add travis build icon

splashx committed:
Fix a typo in README (#2)

Fixes a typo in command level in README.md

jesselang commented:
Nice work on the testing! Thanks for working on that.

I think the test environment can be simplified to leverage travis directly without using docker, and specify `PYTHONPATH` to point to the `testbundle` instead of managing packaging for it. Travis config could look like this:

```
language: python
python:
  - 3.5
  - 3.6
  - nightly

env:
  - PYTHONPATH=tests/testbundle

install:
  - pip3 install .

script:
  - python -m unittest discover -v tests/
```

I've tested this config in a branch, and it seems to work all right. Using this config would make the following files extraneous:

```
.dockerignore
Dockerfile
Makefile
tests/.gitignore
tests/testbundle/.dockerignore
tests/testbundle/.gitignore
tests/testbundle/Dockerfile
tests/testbundle/Makefile
tests/testbundle/setup.py
tests/testbundle/config.yaml
```

splashx committed:
Remove docker and python3.6 dependency (#4)

travis: no docker usage 5fc7101
travis: set maxDiff to None 74d257d
fine-tune cog-command exit, add new tests 80d2fdb
use ImportError instead of ModuleNotFoundError (python3.6 only) - ffbfa5b
change assertion message in test e3808b8

splashx committed:
travis - remove sudo requirement

jesselang commented on .gitignore:
```diff
@@ -1,5 +0,0 @@
-**/*.pyo
-build
-dist
-*.egg-info
```
Can we keep this file as-is? I think it still provides value.

jesselang commented on README.md:
```diff
@@ -27,16 +29,78 @@ directory layout:
+ * `commands\commanda.py` - maps to `!commanda`
+ * `commands\level1\commandb.py` - maps to `!level1-commandb`
+ * `commands\level1\level2a\commandc.py` - maps to `!level1-level2a-commandb`
+ * `commands\level1\level2b\leveln\commandz.py` - maps to `!level1-level2b-lebeln-commandz`
```
Typo: replace `lebeln` with `leveln` please.

jesselang commented on tests/testbundle/Makefile:
```diff
@@ -0,0 +1,23 @@
+docker-shell:
+	docker run --rm -it $(IMAGE_TAG) sh
+
+docker-fresh: docker-clean docker
```
Since the `Dockerfile` is gone here, I'm guessing this file could go away, and the associated `README.md` updated.

jesselang commented on tests/testbundle/README.md:
```diff
@@ -0,0 +1,38 @@
+
+    $ make docker
+
+Requires Python 3.6.x, pip, make, and Docker.
```
If you remove the `Makefile`, the `# Building` section can be removed as well, and the dependency list at the bottom updated.

jesselang commented on README.md:
````diff
@@ -27,16 +29,78 @@ directory layout:
+ 
+
+```
+<bundle_name>
````
It may be clearer to replace `<bundle_name>` here with `<bundle_directory>` since this directory doesn't need to be the same as the bundle name.

jesselang reviewed (changes requested):
Great stuff, Dio! This is nearly ready to merge. Thanks for your efforts here!

- Having the top-level `.gitignore` back would be helpful.
- Found a minor typo, and a suggested edit.
- Curious if `tests/testbundle/Makefile` is needed, especially since the `Dockerfile` was removed.

splashx committed:
typo in README, add gitignore back, fix testbundle README

* typo in README, add gitignore back, fix testbundle README

splashx committed:
typo in README, add gitignore back, fix testbundle README

* typo in README, add gitignore back, fix testbundle README

splashx committed:
remove tests/testbundle/Makefile (#8)

remove tests/testbundle/Makefile
